cosmwasm/packages/std/src/imports.rs

```diff
@@ -187,6 +215,227 @@ impl Api for ExternalApi {
+    ///
+    /// Three Variants are suppported in the input for convenience:
+    ///  - Equal number of messages, signatures, and public keys: Standard, generic functionality.
+    ///  - One message, and an equal number of signatures and public keys: Multiple digital signature
+    /// (multisig) verification of a single message.
+    ///  - One public key, and an equal number of messages and signatures: Verification of multiple
+    /// messages, all signed with the same private key.
+    ///
+    /// Any other variants of input vectors result in an error.
+    ///
+    /// Notes:
+    ///  - The "one-message, with zero signatures and zero public keys" case, is considered the empty
+    /// case.
+    ///  - The "one-public key, with zero messages and zero signatures" case, is considered the empty
+    /// case.
+    ///  - The empty case (no messages, no signatures and no public keys) returns true.
+    fn ed25519_batch_verify(
+        &self,
+        messages: &[&[u8]],
+        signatures: &[&[u8]],
+        public_keys: &[&[u8]],
+    ) -> Result<bool, VerificationError> {
+        let msgs_encoded = encode_sections(messages);
+        let msgs_send = build_region(&msgs_encoded);
+        let msgs_send_ptr = &*msgs_send as *const Region as u32;
+
+        let sigs_encoded = encode_sections(signatures);
+        let sig_sends = build_region(&sigs_encoded);
+        let sigs_send_ptr = &*sig_sends as *const Region as u32;
+
+        let pubkeys_encoded = encode_sections(public_keys);
+        let pubkeys_send = build_region(&pubkeys_encoded);
+        let pubkeys_send_ptr = &*pubkeys_send as *const Region as u32;
+
+        let result =
+            unsafe { ed25519_batch_verify(msgs_send_ptr, sigs_send_ptr, pubkeys_send_ptr) };
+        match result {
+            0 => Ok(true),
+            1 => Ok(false),
+            4 => Err(VerificationError::InvalidSignatureFormat),
+            5 => Err(VerificationError::InvalidPubkeyFormat),
+            10 => Err(VerificationError::GenericErr),
+            error_code => Err(VerificationError::unknown_err(error_code)),
+        }
+    }
+
+    fn secp256k1_sign(&self, message: &[u8], private_key: &[u8]) -> Result<Vec<u8>, SigningError> {
+        let msg_send = build_region(message);
+        let msg_send_ptr = &*msg_send as *const Region as u32;
+        let pk_send = build_region(private_key);
+        let pk_send_ptr = &*pk_send as *const Region as u32;
+
+        let result = unsafe { secp256k1_sign(msg_send_ptr, pk_send_ptr) };
+        let error_code = from_high_half(result);
+        let signature_ptr = from_low_half(result);
+        match error_code {
+            0 => {
+                let signature = unsafe { consume_region(signature_ptr as *mut Region) };
+                Ok(signature)
+            }
+            1000 => Err(SigningError::InvalidPrivateKeyFormat),
+            error_code => Err(SigningError::unknown_err(error_code)),
```

Comment:
Feels like this should be a SigningError::from_i32 function instead of this

Reply:
Just copied this format from CosmWasm v1